MekHQ/src/mekhq/campaign/io/Migration/ForceIconMigrator.java

```diff
@@ -0,0 +1,6891 @@
+/*
+ * Copyright (c) 2021 - The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MekHQ.
+ *
+ * MekHQ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MekHQ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MekHQ. If not, see <http://www.gnu.org/licenses/>.
+ */
+package mekhq.campaign.io.Migration;
+
+import mekhq.MekHQ;
+import mekhq.campaign.icons.ForcePieceIcon;
+import mekhq.campaign.icons.LayeredForceIcon;
+import mekhq.campaign.icons.StandardForceIcon;
+import mekhq.campaign.icons.UnitIcon;
+import mekhq.campaign.icons.enums.LayeredForceIconLayer;
+import org.apache.logging.log4j.LogManager;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * This migrates Force icons from varied sources to Kailan's Pack
+ * This migration occurred in 0.49.6
+ */
+public class ForceIconMigrator {
+    public static StandardForceIcon migrateForceIcon(final StandardForceIcon icon) {
+        if (icon instanceof LayeredForceIcon) {
+            return migrateLayeredForceIcon((LayeredForceIcon) icon);
+        } else if (icon instanceof UnitIcon) {
+            return migrateUnitIcon((UnitIcon) icon);
+        } else if (icon != null) {
+            return migrateStandardForceIcon(icon);
+        } else {
+            LogManager.getLogger().error("Attempted to migrate a null icon, when AbstractIcon is non-nullable by design");
+            return new StandardForceIcon();
+        }
+    }
+
+    //region Layered Force Icon
+    private static LayeredForceIcon migrateLayeredForceIcon(final LayeredForceIcon icon) {
+        final LayeredForceIcon layered = new LayeredForceIcon();
+        migrateAdjustments(icon, layered);
+        migrateAlphanumerics(icon, layered);
+        migrateBackgrounds(icon, layered);
+        migrateFormations(icon, layered);
+        migrateLogos(icon, layered);
+        migrateSpecialModifiers(icon, layered);
+        migrateTypes(icon, layered);
+        return layered;
+    }
+
+    private static void migrateAdjustments(final LayeredForceIcon original, final LayeredForceIcon migrated) {
+        for (final ForcePieceIcon piece : original.getPieces().getOrDefault(LayeredForceIconLayer.ADJUSTMENT, new ArrayList<>())) {
+            switch (piece.getFilename()) {
+                case "Air assault.png":
```

Comment:
That is a very large amount of code to accommodate directory shuffling. I give it about a 0% chance of going through without a large number of  "where did my force icons go" complaints.

Reply:
It's a complete internal pack swapover, not "directory shuffling". Kailan created and organized her pack based on the previous setup while taking advantage of the increased graphical functionality.

If you have a better and/or safer way to migrate the data from the old pack to the new one please feel free to write your own, but I've got faith that this will migrate the vast majority of users without issue, especially given most TO&Es use a small portion of the overall icons available.